### PR TITLE
MinMDNS available policy: libnl

### LINKED ISSUE #22046: Linux address iteration uses deprecated/temporary and other generally "no usable" Ip addresses
#### Problem
The API that linux uses for fetching IP addresses is not aware of interface flags.

As a result, addresses marked temporary (before Duplicate Address Detection/DAD) or deprecated addresses (from old routing advertisements) are still used and reported as DNSSD 'usable addresses'

 #### Proposed Solution
 Figure out a way to better filter usable addresses.
 
 One step would be to have a linux-specific platform implementation that is more aware of address flags (e.g. using libnl).
 Another step would be to allow for hooks in DNSSD (minmdns specifically) for better integration for applications to control IP address logic, like what interface to advertise on and which IP address(es) to allow for such advertisement.

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem

Add an example minmdns policy using libnl (linux specific). This improves on existing policies by
better flag detection: loopback interface detection, various address detection (temporary/deprecated/ ...)

#### Change overview
LibNL policy implementation and compile flag (does NOT change defaults)

Fixes #22046 

#### Testing
Manually tested minmdns examples locally.

### CODE REVIEW

**Comment by andy31415** [comment]
Postponed past 1.0: applications can define their own policy as needed, a built-in policy is a convenience but not strictly 1.0 required.

File: examples/minimal-mdns/client.cpp [hunk 1]
```diff
@@ -22,8 +22,8 @@
 
 #include <inet/InetInterface.h>
 #include <inet/UDPEndPoint.h>
+#include <lib/dnssd/MinimalMdnsServer.h>
```
Comment (andy31415) [inline examples/minimal-mdns/client.cpp:25]: comment to prevent merging: wait until after 1.0 branch

**Review by bzbarsky-apple — APPROVED** [review]

File: src/lib/dnssd/MinimalMdnsServer.cpp [hunk 2]
```diff
@@ -22,8 +22,16 @@
 #define CHIP_MINMDNS_DEFAULT_POLICY 0
 #endif
 
+#ifndef CHIP_MINMDNS_LIBNL_POLICY
+#define CHIP_MINMDNS_LIBNL_POLICY 0
+#endif
+
 #if CHIP_MINMDNS_DEFAULT_POLICY
-#include <lib/dnssd/minimal_mdns/AddressPolicy_DefaultImpl.h>
+#include <lib/dnssd/minimal_mdns/AddressPolicy_DefaultImpl.h> // nogncheck
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/lib/dnssd/MinimalMdnsServer.cpp:30]: I believe gn rules enforce that (in theory). Not sure if I can static assert more.

**Comment by andy31415** [comment]
Switching to accepted due to it not changing default behavior and being platform-specific. Feedback from @tcarmelveilleux was that this is valuable and desired for 1.0

File: scripts/build/builders/host.py [hunk 3]
```diff
@@ -280,6 +281,11 @@ def __init__(self, root, runner, app: HostApp, board=HostBoard.NATIVE,
                 # so setting clang is not correct
                 raise Exception('Fake host board is always gcc (not clang)')
 
+        if minmdns_address_policy:
+            if use_platform_mdns:
+                raise Exception('Address policy applies to minmdns only')
```
[the remark this answers is not in the document]
Comment (andy31415) [reply scripts/build/builders/host.py:286]: Will look at a followup. I believe we do not have too much precedence on this kind of checks - most flags are translated into compile macros without double-checking if they make sense (e.g. I believe one can enable thread options for wifi builds and such and they will be a noop, like in this case)

**Pull request merged**